src/nexus.rs

```diff
@@ -0,0 +1,246 @@
+use futures::{stream, StreamExt};
+use reqwest_middleware::ClientWithMiddleware;
+use serde_json::Value;
+use url::Url;
+
+use crate::client::get_nexus_client;
+use crate::errors::NexusError;
+use crate::gem::Gemspec;
+
+const CONCURRENT_REQUESTS: usize = 3;
+
+type ResultCollection = Vec<Result<NexusResult, NexusError>>;
+
+///
+/// Check packages existance in Nexus repository
+///
+pub(crate) async fn check_packages(
+    packages: &Vec<Gemspec>,
+    nexus_url: &str,
+    verbose: bool,
+) -> Result<Vec<NexusResult>, NexusError> {
+    let nexus = Nexus::new(nexus_url)?;
+
+    let nexus_results = stream::iter(packages)
+        .map(|package| async { nexus.check_package(package).await })
+        .buffer_unordered(CONCURRENT_REQUESTS)
+        .collect::<ResultCollection>()
+        .await;
+
+    let (oks, errors): (ResultCollection, ResultCollection) =
+        nexus_results.into_iter().partition(Result::is_ok);
+    let oks: Vec<NexusResult> = oks.into_iter().map(Result::unwrap).collect();
+
+    if verbose {
+        errors
+            .into_iter()
+            .map(Result::unwrap_err)
+            .for_each(|error| println!("{}", error));
+    }
+
+    Ok(oks)
+}
+
+pub(crate) struct Nexus {
+    // Nexus repository url
+    repo_url: Url,
+
+    // Which type of artefact should to request
+    format_artefact: String,
+
+    // Nexus client instance
+    client: ClientWithMiddleware,
+}
+
+#[derive(Debug)]
+pub(crate) struct NexusResult {
+    pub(crate) name: String,
+    pub(crate) version: String,
+    pub(crate) purl: String,
+    pub(crate) is_exist: bool,
+}
+
+impl Nexus {
+    ///
+    /// Initializes new Nexus instance
+    ///
+    pub(crate) fn new(repo_url: &str) -> Result<Self, NexusError> {
+        let repo_url =
+            Url::parse(repo_url).map_err(|_| NexusError::UrlParse(repo_url.to_string()))?;
+        let client = get_nexus_client().map_err(|_| NexusError::BuildClient)?;
+
+        Ok(Nexus {
+            format_artefact: String::from("rubygems"),
+            repo_url,
+            client,
+        })
+    }
+
+    ///
+    /// Check package existance in Nexus repository
+    ///
+    pub(crate) async fn check_package(&self, package: &Gemspec) -> Result<NexusResult, NexusError> {
+        let name = &package.name;
+        let version = &package.version;
+        let response = self.send_request(name, version).await;
+
+        let check_result = self.check_response(response);
+
+        check_result.map(|is_exist| NexusResult {
+            name: name.to_string(),
+            version: version.to_string(),
+            purl: package.purl.to_string(),
+            is_exist,
+        })
+    }
+
+    //
+    // Parses respose json and try to check whether package exists in
+    // Nexus repository
+    //
+    fn check_response(&self, response: Result<Value, NexusError>) -> Result<bool, NexusError> {
+        response.map(|json| !json["items"].as_array().unwrap().is_empty())
```

Comment:
Maybe should to add explicit error instead `unwrap`?

Reply:
No it's not `Result`, it's `Option`. Better panics if recevied wrong response structure and fix it